src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+    ), HashF64Context, std.hash_map.default_max_load_percentage),
+    compressed_values: *ArrayList(u8),
+) !void {
+    // Iterate over the hash to hash map.
+    var hash_to_hash_iterator = merged_segments_metadata_map.iterator();
+    while (hash_to_hash_iterator.next()) |hash_to_hash_entry| {
+        const current_interception: f64 = hash_to_hash_entry.key_ptr.*;
+        // Append the `current_interception`.
+        try appendValue(f64, current_interception, compressed_values);
+
+        // Append the number of slopes hashed by the `current_interception`.
+        try appendValue(usize, hash_to_hash_entry.value_ptr.*.count(), compressed_values);
+        var hash_to_array_iterator = hash_to_hash_entry.value_ptr.*.iterator();
+
+        while (hash_to_array_iterator.next()) |hash_to_array_entry| {
+            const current_slope = hash_to_array_entry.key_ptr.*;
+            // Append the `current_slope` that approximates the segment.
+            try appendValue(f64, current_slope, compressed_values);
+
+            // Append the number of timestamps that the `current_slope` is hashing.
+            try appendValue(usize, hash_to_array_entry.value_ptr.*.items.len, compressed_values);
+            var previous_timestamp: usize = 0;
+            // Iterate over the timestamps and append them.
+            for (hash_to_array_entry.value_ptr.*.items) |timestamp| {
+                try appendValue(usize, timestamp - previous_timestamp, compressed_values);
+                previous_timestamp = timestamp;
+            }
+        }
+    }
+}
+
+/// Quantizes the given `value` by the specified `error_bound`. This process ensures that
+/// the quantized value remains within the error bound of the original value. If the
+/// `error_bound` is equal to zero, the value is directly returned.
+fn quantize(value: f64, error_bound: f32) f64 {
+    if (error_bound != 0)
+        return @floor(value / error_bound) * error_bound;
+    return value;
+}
+
+/// Appends the `metadata` to the hash map `metadata_map`. The `allocator` to use for allocating
+/// the memory for a new ArrayList if the `metadata.interception` does not exist.
+fn appendSegmentMetadata(
+    metadata_map: *HashMap(
+        f64,
+        ArrayList(SegmentMetadata),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ),
+    metadata: SegmentMetadata,
+    allocator: mem.Allocator,
+) !void {
+    const get_result = try metadata_map.getOrPut(metadata.interception);
+    if (!get_result.found_existing) {
+        get_result.value_ptr.* = ArrayList(SegmentMetadata).init(allocator);
+    }
+    try get_result.value_ptr.*.append(metadata);
+}
+
+/// Returns a comparator function that compares SegmentMetadata by `lower_bound_slope`.
+fn compareMetadataBySlope() fn (void, SegmentMetadata, SegmentMetadata) bool {
+    return struct {
+        pub fn inner(_: void, metadata_one: SegmentMetadata, metadata_two: SegmentMetadata) bool {
```

Comment:
Unclear why this is not just a private function in the module scope instead of being part of a struct and then being returned as a function without the struct.

Reply:
Great. Fixed!

Reply:
Very minor, but maybe remove "Returns a" from the comment as the function now is the comparator and returns a `bool`.

Reply:
Fixed.